ДЗ 15 спринта, 19 когорта андроид разработчик

powerrewop opened the pull request:
Исправлены замечания
1) полный рефакторинг всех usecase / интеракторов. Сделал в соответствии примеров в 15 спринте
2) переделал Creator в соответствии примеров в 15 спринте
3) избавился от передачи контекста в коде



ДЗ 15 спринта (чистая архитектура), 19 когорта андроид разработчик (перешел из 17 когорты)

powerrewop committed:
CA settings finish

powerrewop committed:
CA settings, main finish

powerrewop committed:
CA player, settings, main finish

powerrewop committed:
CA all

kekulta commented on app/src/main/java/com/practicum/playlistmaker/creator/PlayerUseCaseCreator.kt:
```diff
@@ -0,0 +1,23 @@
+import com.practicum.playlistmaker.domain.usecase.PlayerControlUseCase
+import com.practicum.playlistmaker.presentation.OptionsPlayerActivity
+
+class PlayerUseCaseCreator(private val act: Activity) {
```
В Creator нельзя передавать Activity Context, это приведёт к утечкам памяти.
Более правильным подходом будет сделать поле для контекста и инициализировать его с помощию Application Context из класса App.
Application Context имеет время жизни такое же как время жизни приложение, так что утечка будет невозможна.
Затем использовать этот контекст везде, где он нужен, e.g. создание баз данных, навигация, etc.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/creator/PlayerUseCaseCreator.kt:
```diff
@@ -0,0 +1,23 @@
+        val playerControl = PlayerControlImpl()
+        val playerControlUseCase = PlayerControlUseCase(playerControl)
+        return CreatorPlayerControlModel(playerControl, playerControlUseCase)
+    }
```
Нет никакого смысла разбиавать Creator на несколько классов и создавать классы холдеры для него.
Лучше сделать один `object Creator` и там уже сделать пачку методов, которые будут возвращать что нужно. Если какие-то из методов используются только внутри Creator, например, репозитории используются только для создания интеракторов, то эти методы можно сделать приватными.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/data/model/TrackDto.kt:
```diff
@@ -1,16 +1,16 @@
 
-data class Track(
+class TrackDto (
     var trackName: String,
     var artistName: String,
     @SerializedName("trackTimeMillis") var trackTime: Int,
     var artworkUrl100: String?,
     var trackId: Long,
     var isHistory: Boolean = false,
     var collectionName: String?,
-    var releaseDate: String,
+    var releaseDate: String?,
     var primaryGenreName: String,
     var country: String,
```
Мы никогда не доверяем бекенду. Там может случится баг или произойти изменения 
в API и какое-то из этих полей не придёт. Тогда приложение просто упадёт  и мы 
никак не сможем это корректно обработать.
Правильный подход будет сделать все поля в DTO нуллабельными и потом мапить 
это в Domain модель, где поля уже не нуллабельные. 
Могут быть разные подходы к обработке нуллабельных полей. Можно, например, 
если какого-то поля не хватает, то просто залоггировать это и пропустить этот 
объект. Можно заполнять недостующие поля какими-то заглушками и показывать их. 
Или можно заполнять их специальными значениями при виде которых UI будет 
просто скрывать эти поля.
Гланый плюс, что мы избежим падения приложения в случае ошибки бекенда.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/activitys/SearchActivityNew.kt:
```diff
@@ -0,0 +1,165 @@
+    private var trAdapt: TrackAdapter? = null
+
+    private lateinit var userTextWorkImpl: UserTextWorkImpl
+    private lateinit var userTextWorkUseCase: UserTextWorkUseCase
```
Лейтинитов стоит по возможности избегать. Это окей использовать их для вьюшек, пока вы не перешли на view binding, но в остальных случаях их лучше не использовать.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/activitys/SearchActivityNew.kt:
```diff
@@ -0,0 +1,165 @@
+        val getHistorySearchUseCase = searchUseCaseCreator.getHistorySearchUseCase()
+        /////////////////////////
+
+        val optionsSearchActivity = OptionsSearchActivity(
+            trAdapt,
+            recycler,
+            problemLayout,
+            problemImage,
+            problemText,
+            buttonUpdate,
+            historyText,
+            buttonHistoryClear,
+            inputEditText,
+            clearButton,
+            userText,
+            this,
+            layoutProgressBar
+        )
```
Плохая идея передвать куда-то вьюшки. Весь этот код должен быть в активити. В этом весь смысл существования активити, работать с вьюшками.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/activitys/SearchActivityNew.kt:
```diff
@@ -0,0 +1,165 @@
+            layoutProgressBar.isVisible = true
+        }
+
+        buttonHistoryClear!!.setOnClickListener {
+            clearHistoryUseCase.clearHistory()
+            optionsSearchActivity.hideHistoryElements()
+        }
```
Нужно по возможности избегать `!!`, это очень опасное поведение и может привести к крашу.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/OptionsSearchActivity.kt:
```diff
@@ -0,0 +1,171 @@
+        layoutProgressBar.isVisible = false
+        if (!hsActive) {
+
+            if ((resultCode == 200) && ((tracks == null) || (tracks.isEmpty()))) {
```
Можно заменить на isNullOrEmpty.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/OptionsSearchActivity.kt:
```diff
@@ -0,0 +1,171 @@
+import com.practicum.playlistmaker.presentation.activitys.PlayerActivity
+
+class OptionsSearchActivity(
+    private var trAdapt: TrackAdapter?,
+    private var recycler: RecyclerView?,
+    private val problemLayout: LinearLayout?,
+    private val problemImage: ImageView,
+    private val problemText: TextView,
+    private val buttonUpdate: Button,
+    private val historyText: TextView?,
+    private val buttonHistoryClear: Button?,
+    private val inputEditText: EditText?,
+    private val clearButton: ImageView,
+    private var userText: String,
+    private val act: Activity,
+    private val layoutProgressBar: LinearLayout
```
Непонятно зачем делать параметры нуллабл и затем использовать `!!`

powerrewop committed:
Исправлены замечания
1) скорректировано поведение прогресс бара
2) удален usecase содержащий тип Intent и все что с ним было связано, код с интентами вернул в активити
3) увеличил размер картинки альбома в форме плейера

powerrewop committed:
Исправлены замечания
1) скорректировано поведение прогресс бара
2) удален usecase содержащий тип Intent и все что с ним было связано, код с интентами вернул в активити
3) увеличил размер картинки альбома в форме плейера

powerrewop committed:
исправление замечаний

kekulta commented on app/src/main/java/com/practicum/playlistmaker/data/storage/MediaPlayerNew.kt:
```diff
@@ -0,0 +1,89 @@
+    private var playerState = STATE_DEFAULT
+    private var mainThreadHandler = Handler(Looper.getMainLooper())
+
+    lateinit var setPlImage: () -> Unit
+    lateinit var setPaImage: () -> Unit
+    lateinit var setTime: (timePlay: String) -> Unit
```
lateinit лучше не использовать лишний раз. Если что-то может инициализироваться в конструкторе, а не в `onCreate`, то оно должно инициализироваться там.

lateinit это костыль. Он ломает логику языка и делает его менее безопасным.
Добавлена эта штука для интеропа с Джавой, например, некоторые функции в фреймворке Spring не работали без этого.

Если где-то есть лейтинит, то это повод очень сильно задуматься.
Почти всегда окажется, что он там либо вообще не нужен, либо можно переписать с использованием делегата `by lazy`.
Если совсем никак, то nullable предпочительнее, т.к. заставляет следить за руками.
В каких-то случаях лейтинит оправдан, например, переменные для View, если не используешь ViewBinding.

Так же сам ViewBinding можно сделать в lateinit, но предпочтительнее использовать конструкцию вроде:
```
private val _binding: ActvityBinding? = null
private val binding: ActvityBinding  get() = requireNotNull(_binding) { "Binding wasn't initiliazed!" }
```

Это нужно, чтобы ошибка была более чёткой.
Так же, например, если использовать binding внутри фрагментов, то переменную нужно ставить null в `onViewDestroyed()` иначе возможна утечка памяти.
При использовании Binding внутри Actvity это не обязательно.

В новом Котлине появилась фича `Backing Fields`, которая сделает эту конутрукцию менее многословной, но она пока что в бете.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/data/storage/MediaPlayerNew.kt:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    companion object {
+        private const val STATE_DEFAULT = 0
+        private const val STATE_PREPARED = 1
+        private const val STATE_PLAYING = 2
+        private const val STATE_PAUSED = 3
+        private const val TIMER_DELAY = 250L
+    }
```
Стейт удобнее моделировать с помощью sealed class/interface.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/OptionsSearchActivity.kt:
```diff
@@ -0,0 +1,156 @@
+import com.practicum.playlistmaker.domain.model.Track
+
+class OptionsSearchActivity(
+    private var trAdapt: TrackAdapter?,
+    private var recycler: RecyclerView?,
+    private val problemLayout: LinearLayout?,
```
Нет никакого смысла делать тут nullable и var.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/domain/usecase/MediaplayerUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+import com.practicum.playlistmaker.domain.storage.interfaces.MediaPlayerRepository
+
+class MediaplayerUseCase(private val mediaPlayerRepository: MediaPlayerRepository) {
+     fun preparePlayer(url: String?, setPlayImage: () -> Unit, setPauseImage: () -> Unit, setTimePlay: (timePlay: String) -> Unit) {
```
Это скорее коллбэки, которые можно назвать `onPause`, `onStop`, etc.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/presentation/OptionsSearchActivity.kt:
```diff
@@ -0,0 +1,156 @@
+        recycler?.isVisible = true
+        layoutProgressBar.isVisible = false
+    }
+    @SuppressLint("SuspiciousIndentation")
+    fun setFocus(hasFocus: Boolean, tracks: List<Track>?){
+        layoutProgressBar.isVisible = false
+        var historyActive = false;
+            if (hasFocus && inputEditText!!.text.isEmpty()) {
+                historyActive = true
+            }else{
+                historyActive = false
+            }
+
+        if(inputEditText!!.text.isEmpty()){
+            hsActive = true
+        }else
+        {
+            hsActive = false
+        }
+
+        showHistory(historyActive, tracks)
+
+    }
```
Очень много кода нужно порефакторить для большей читаемости и соответвия хорошим практикам.
Просто для примера я отрефакторил эту функцию `setFocus()`. Единственное, что я поменял за её пределами это убрал `?` из `inputEditText` т.к. там не нужна нуллабельность:
```kotlin
fun setFocus(hasFocus: Boolean, tracks: List<Track>?) {
    layoutProgressBar.isVisible = false
    //Вероятно historyActive и hsActive должны быть одной переменной.
    hsActive = inputEditText.text.isEmpty()
    showHistory(hasFocus && hsActive, tracks)
}
```

Мест, где можео и нужно провести подобный рефакторинг ещё очень много в проекте.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/creator/Creator.kt:
```diff
@@ -0,0 +1,91 @@
+import com.practicum.playlistmaker.domain.usecase.ParamDataUseCase
+
+object Creator {
+    private fun getAppTheme(): AppTheme{
+        return AppTheme()
+    }
+    private fun getAppThemeRepository(): AppThemeRepository{
+        return AppThemeRepositoryImpl(AppTheme())
```
Не используешь `getAppTheme()` и ещё пару функций в креаторе, хотя объявляешь их.
